lib/livebook_web/helpers.ex

```diff
@@ -82,4 +86,45 @@ defmodule LivebookWeb.Helpers do
   def format_datetime_relatively(date) do
     date |> DateTime.to_naive() |> Livebook.Utils.Time.time_ago_in_words()
   end
+
+  @doc """
+  Enhance the output_panel data with outputs and input_views.
+  """
+  @spec enrich_output_panel_data(Data.t()) :: map()
+  def enrich_output_panel_data(data) do
+    changed_input_ids = Data.changed_input_ids(data)
+
+    data.notebook.output_panel
+    |> Map.from_struct()
+    |> update_in(
+      [
+        Access.key(:rows),
+        Access.all(),
+        Access.key(:items),
+        Access.all()
+      ],
+      fn item ->
+        {:ok, cell, _section} = Notebook.fetch_cell_and_section(data.notebook, item.cell_id)
+
+        item
+        |> Map.put(:outputs, cell.outputs)
+        # TODO fix for multiple outputs
+        |> Map.put(
+          :input_views,
+          input_views_for_output(cell.outputs |> hd(), data, changed_input_ids)
+        )
```

Comment:
I still struggle with this. From [main](https://github.com/livebook-dev/livebook/blob/32a98ff0af5f97586b377609c7366b209522a317/lib/livebook_web/live/app_session_live.ex#L450-L459)
```elixir
defp visible_outputs(notebook) do
    for section <- Enum.reverse(notebook.sections),
        cell <- Enum.reverse(section.cells),
        Cell.evaluable?(cell),
        output <- filter_outputs(cell.outputs, notebook.app_settings.output_type),
        do: {cell.id, output}
  end

  defp filter_outputs(outputs, :all), do: outputs
  defp filter_outputs(outputs, :rich), do: rich_outputs(outputs)
```
Is this correct that for output_type :all the filter_outputs returns the list of output and saves it in output?

Reply:
`:all`/`:rich` goes through all outputs in the notebook and filters which ones we want to show in the app. In case of `:all` we keep all outputs, and note that `output <- filter_outputs(...)` just iterates over the outputs. I may not be getting the question, let me know :)

Reply:
Ahh shoot I got it. Thank you, forgot it's a loop :see_no_evil: